res/controllers/midi-components-0.0.js

```diff
@@ -301,8 +309,48 @@
         shift: function () {
             this.inKey = 'hotcue_' + this.number + '_clear';
         },
+        getColor: function() {
+            if (this.colorIdKey !== undefined) {
+                return color.predefinedColorFromId(engine.getValue(this.group,this.colorIdKey));
+            } else {
+                return false;
+            }
+        },
+        output: function(value) {
+            var outval = this.outValueScale(value);
+            if (this.colorIdKey !== undefined && outval !== this.off) {
+                this.outputColor(engine.getValue(this.group, this.colorIdKey));
+            } else {
+                this.send(outval);
+            }
+        },
+        outputColor: function (id) {
+            var color = this.colors[id];
+            if (color instanceof Array) {
+                if (color.length !== 3) {
+                    print("ERROR: invalid color array for id: " + id);
+                    return;
+                }
+                if (this.sendRGB === undefined) {
+                    print("ERROR: no custom method for sending rgb defined!");
+                    return;
+                }
+                this.sendRGB(color);
+            } else if (typeof color === 'number') {
+                this.send(color);
+            }
+        },
+        connect: function() {
+            Button.prototype.connect.call(this); // call parent connect
+            if (undefined !== this.group && this.colorIdKey !== undefined) {
+                this.connections.push(engine.makeConnection(this.group, this.colorIdKey, function (id) {
```

Comment:
No actually do want multiple callbacks (which is also why I am calling the parent connect).
The color_id key is only responsible for the color of a hotcue but it doesn't actually tell me if the hotcue is unset or just doesn't have a color assigned. So there is a connection that controls if the hotcue is actually assigned (the one that gets made by the parent connect) and a second connection that tells me the color.

Reply:
Okay, thank you for clarifying. In that case, I think it would be better to assign to `this.connections[1]` than push onto the array. If `connect` somehow gets called before `disconnect`, that could create a situation where `this.connections` has more than two members, which could get confusing when debugging. Also, please add a comment explaining that `this.connections[0]` is created by calling Button.prototype.connect.

Reply:
> In that case, I think it would be better to assign to `this.connections[1]` than push onto the array. If `connect` somehow gets called before `disconnect`, that could create a situation where `this.connections` has more than two members, which could get confusing when debugging.

If `disconnect` is called first and `connect` is called afterwards, `this.connections` will already have 3 members, because `disconnect` does not remove the connection from the array, just disconnects it (i.e. `this.connections` contains 2 connected connections and 1 disconnected connection). This also means that after calling `disconnect` and `connect`, calling `trigger` will invoke the callback twice because trigger on a `disconnected` connection apparently still invokes the callback.

TL;DR: Better use `this.connections[1] = ...` instead of `this.connections.push(...)`.